Build failure on GNOME continuous integration pipeline

p11-kit fails to build with this error:

```
FAILED: p11-kit/p11-kit-server-testable 
ccache cc  -o p11-kit/p11-kit-server-testable 'p11-kit/077403d@@p11-kit-server-testable@exe/server.c.o' -Wl,--no-undefined -Wl,--as-needed -O2 -g -m64 -mtune=generic -Wl,--start-group p11-kit/libp11-kit-testable.a common/libp11-common.a common/libp11-library.a common/libp11-tool.a /usr/lib/libsystemd.so /usr/lib/libffi.so -ldl -Wl,--end-group '-Wl,-rpath,$ORIGIN/:$ORIGIN/../common' -Wl,-rpath-link,/ostbuild/source/p11-kit/_build/p11-kit:/ostbuild/source/p11-kit/_build/common 
common/libp11-library.a(library.c.o): In function `p11_library_init_impl':
/ostbuild/source/p11-kit/_build/../common/library.c:163: undefined reference to `pthread_atfork'
collect2: error: ld returned 1 exit status
```

Using the Meson build. The common library does not depend on pthreads, even if it uses pthreads API.

--- a/common/meson.build
+++ b/common/meson.build
@@ -31,7 +31,8 @@ libp11_library = static_library('p11-library', 'library.c',
 
 libp11_library_dep = declare_dependency(include_directories: [configinc,
                                                               commoninc],
-                                        link_with: libp11_library)
+                                        link_with: libp11_library,
+                                        dependencies: thread_deps)
 
 libp11_test_sources = [
   'mock.c',
@@ -42,7 +43,8 @@ libp11_test = static_library('p11-test', libp11_test_sources,
                              include_directories: configinc)
 libp11_test_dep = declare_dependency(include_directories: [configinc,
                                                            commoninc],
-                                     link_with: libp11_test)
+                                     link_with: libp11_test,
+                                     dependencies: thread_deps)
 
 libp11_tool_sources = [
   'tool.c'

--- a/common/meson.build
+++ b/common/meson.build
@@ -29,10 +29,8 @@ libp11_library = static_library('p11-library', 'library.c',
                                 gnu_symbol_visibility: 'hidden',
                                 include_directories: configinc)
 
-libp11_library_dep = declare_dependency(include_directories: [configinc,
-                                                              commoninc],
-                                        link_with: libp11_library,
-                                        dependencies: thread_deps)
+libp11_library_dep = declare_dependency(link_with: libp11_library,
+                                        dependencies: [libp11_common_dep] + thread_deps)
 
 libp11_test_sources = [
   'mock.c',
@@ -41,10 +39,8 @@ libp11_test_sources = [
 
 libp11_test = static_library('p11-test', libp11_test_sources,
                              include_directories: configinc)
-libp11_test_dep = declare_dependency(include_directories: [configinc,
-                                                           commoninc],
-                                     link_with: libp11_test,
-                                     dependencies: thread_deps)
+libp11_test_dep = declare_dependency(link_with: libp11_test,
+                                     dependencies: [libp11_common_dep] + thread_deps)
 
 libp11_tool_sources = [
   'tool.c'
@@ -56,9 +52,8 @@ endif
 
 libp11_tool = static_library('p11-tool', libp11_tool_sources,
                              include_directories: configinc)
-libp11_tool_dep = declare_dependency(include_directories: [configinc,
-                                                           commoninc],
-                                     link_with: libp11_tool)
+libp11_tool_dep = declare_dependency(link_with: libp11_tool,
+                                     dependencies: [libp11_common_dep])
 
 # Tests ----------------------------------------------------------------
 

--- a/meson.build
+++ b/meson.build
@@ -75,19 +75,15 @@ tests_c_args = [
 
 conf.set('SIZEOF_UNSIGNED_LONG', cc.sizeof('unsigned long'))
 
-pthread_create_deps = []
 nanosleep_deps = []
 dlopen_deps = []
 socket_deps = []
+thread_deps = []
 
 if host_system != 'windows'
-  if not cc.has_function('pthread_create')
-    libpthread = cc.find_library('pthread', required: false)
-    if cc.has_function('pthread_create', dependencies: libpthread)
-      pthread_create_deps += libpthread
-    else
-      error('could not find pthread_create')
-    endif
+  thread_deps += dependency('threads')
+  if not cc.has_function('pthread_create', dependencies: thread_deps)
+    error('could not find pthread_create')
   endif
 
   if not cc.has_function('nanosleep')

--- a/p11-kit/meson.build
+++ b/p11-kit/meson.build
@@ -213,7 +213,7 @@ foreach name : p11_kit_tests
   t = executable(name, '@0@.c'.format(name),
                  c_args: tests_c_args + libp11_kit_testable_c_args,
                  include_directories: [configinc, commoninc],
-                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps + pthread_create_deps,
+                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
                  link_with: libp11_kit_testable)
   test(name, t)
 endforeach
@@ -227,7 +227,7 @@ foreach name : p11_kit_progs
   t = executable(name, '@0@.c'.format(name),
                  c_args: tests_c_args,
                  include_directories: [configinc, commoninc],
-                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps + pthread_create_deps,
+                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
                  link_with: libp11_kit_testable)
 endforeach
 

--- a/p11-kit/meson.build
+++ b/p11-kit/meson.build
@@ -34,8 +34,7 @@ libp11_kit_internal_c_args = [
 
 libp11_kit_internal = static_library('p11-kit-internal',
                                      libp11_kit_internal_sources,
-                                     dependencies: [libp11_common_dep,
-                                                    libp11_library_dep] + libffi_deps,
+                                     dependencies: [libp11_library_dep] + libffi_deps,
                                      c_args: libp11_kit_internal_c_args,
                                      implicit_include_directories: false,
                                      gnu_symbol_visibility: 'hidden')
@@ -117,7 +116,7 @@ libp11_kit_testable = static_library('p11-kit-testable',
                                      libp11_kit_sources,
                                      c_args: libp11_kit_testable_c_args,
                                      implicit_include_directories: false,
-                                     dependencies: [libp11_common_dep, libp11_library_dep] + libffi_deps + dlopen_deps)
+                                     dependencies: [libp11_library_dep] + libffi_deps + dlopen_deps)
 
 p11_kit_sources = [
   'lists.c',
@@ -127,14 +126,14 @@ p11_kit_sources = [
 executable('p11-kit',
            p11_kit_sources,
            c_args: common_c_args,
-           dependencies: [libp11_tool_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
+           dependencies: [libp11_tool_dep] + libffi_deps + dlopen_deps,
            link_with: libp11_kit,
            install: true)
 
 executable('p11-kit-remote',
            'remote.c',
            c_args: common_c_args,
-           dependencies: [libp11_tool_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
+           dependencies: [libp11_tool_dep] + libffi_deps + dlopen_deps,
            link_with: libp11_kit,
            install: true,
            install_dir: prefix / privatedir)
@@ -142,7 +141,7 @@ executable('p11-kit-remote',
 executable('p11-kit-remote-testable',
            'remote.c',
            c_args: common_c_args,
-           dependencies: [libp11_tool_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
+           dependencies: [libp11_tool_dep] + libffi_deps + dlopen_deps,
            link_with: libp11_kit_testable)
 
 executable('p11-kit-server',
@@ -150,7 +149,7 @@ executable('p11-kit-server',
            c_args: common_c_args + [
              '-DP11_KIT_REMOTE="p11-kit-remote"'
            ],
-           dependencies: [libp11_tool_dep, libp11_common_dep] + libsystemd_deps + libffi_deps + dlopen_deps,
+           dependencies: [libp11_tool_dep] + libsystemd_deps + libffi_deps + dlopen_deps,
            implicit_include_directories: false,
            link_with: libp11_kit,
            install: true,
@@ -162,7 +161,7 @@ executable('p11-kit-server-testable',
              '-DP11_KIT_REMOTE="p11-kit-remote-testable"'
            ],
            implicit_include_directories: false,
-           dependencies: [libp11_tool_dep, libp11_common_dep] + libsystemd_deps + libffi_deps + dlopen_deps,
+           dependencies: [libp11_tool_dep] + libsystemd_deps + libffi_deps + dlopen_deps,
            link_with: libp11_kit_testable)
 
 # Check if compilation succeeds with CRYPTOKI_GNU=1
@@ -213,7 +212,7 @@ foreach name : p11_kit_tests
   t = executable(name, '@0@.c'.format(name),
                  c_args: tests_c_args + libp11_kit_testable_c_args,
                  include_directories: [configinc, commoninc],
-                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
+                 dependencies: [libp11_test_dep] + libffi_deps + dlopen_deps,
                  link_with: libp11_kit_testable)
   test(name, t)
 endforeach
@@ -227,7 +226,7 @@ foreach name : p11_kit_progs
   t = executable(name, '@0@.c'.format(name),
                  c_args: tests_c_args,
                  include_directories: [configinc, commoninc],
-                 dependencies: [libp11_test_dep, libp11_common_dep] + libffi_deps + dlopen_deps,
+                 dependencies: [libp11_test_dep] + libffi_deps + dlopen_deps,
                  link_with: libp11_kit_testable)
 endforeach
 
@@ -264,8 +263,7 @@ foreach name, sources : mock_sources
   shared_module(name,
                 sources,
                 name_prefix: '',
-                include_directories: [configinc, commoninc],
-                dependencies: [libp11_test_dep, libp11_common_dep])
+                dependencies: [libp11_test_dep])
 endforeach
 
 p11_kit_pc_variables = [

--- a/trust/meson.build
+++ b/trust/meson.build
@@ -49,8 +49,7 @@ shared_module('p11-kit-trust',
               'module-init.c',
               name_prefix: '',
 	      c_args: p11_kit_trust_c_args,
-              dependencies: [libp11_library_dep,
-                             libp11_common_dep] + libtasn1_deps,
+              dependencies: [libp11_library_dep] + libtasn1_deps,
               link_with: libtrust_data,
               install: true,
               install_dir: prefix / p11_module_path)
@@ -90,7 +89,6 @@ executable('trust',
            c_args: common_c_args,
            link_with: libtrust_data,
            dependencies: [libp11_kit_dep,
-                          libp11_common_dep,
                           libp11_tool_dep] + libffi_deps + dlopen_deps + libtasn1_deps,
            install: true)
 
@@ -133,8 +131,7 @@ foreach name : trust_tests
                  c_args: common_c_args + tests_c_args + libtrust_testable_c_args,
                  dependencies: [libp11_kit_dep,
                                 libp11_library_dep,
-                                libp11_test_dep,
-                                libp11_common_dep] + dlopen_deps,
+                                libp11_test_dep] + dlopen_deps,
                  link_with: [libtrust_testable, libtrust_data, libtrust_test])
   test(name, t)
 endforeach
@@ -156,7 +153,6 @@ foreach name : trust_progs
                  c_args: tests_c_args,
                  dependencies: [libp11_kit_dep,
                                 libp11_library_dep,
-                                libp11_test_dep,
-                                libp11_common_dep] + libffi_deps + dlopen_deps,
+                                libp11_test_dep] + libffi_deps + dlopen_deps,
                  link_with: [libtrust_testable, libtrust_data, libtrust_test])
 endforeach

